spec/MultiPoolJwtDecoder.spec.php

```diff
@@ -0,0 +1,139 @@
+        
+        allow($this->verifier)->toReceive('verify')->andRun(function ($token) {
+            return $token;
+        });
+
+        allow($this->verifier)->toReceive('getUserPool')->andReturn($this->userPool);
+
+        $this->decoder = new MultiPoolJwtDecoder([$this->verifier]);
+
+        // Give it a leeway of 100 years to allow testing decoding the token without exception
+        JWT::$leeway = 60 * 60 * 24 * 365 * 100;
+    });
+
+    afterEach(function () {
+        JWT::$leeway = 60;
+    });
+
+    describe('->decode()', function () {
+        it('should reject invalid structure', function () {
+            $sut = function () {
+                $invalid = '11111.222222';
+                $this->decoder->decode($invalid, $this->validJwk);
+            };
+            expect($sut)->toThrow(new InvalidStructureException());
+        });
+
+        it('should reject a structure that contains non-base64 encoded items', function () {
+            $sut = function () {
+                $invalid = 'as%iudfh9w=8uihf.asi%udfh9w=8uihf.as%iudfh9w=8uihf';
+                $this->decoder->decode($invalid, $this->validJwk);
+            };
+            expect($sut)->toThrow(new InvalidStructureException());
+        });
+
+        it('should reject an empty structure', function () {
+            $sut = function () {
+                $invalid = '..';
+                $this->decoder->decode($invalid, $this->validJwk);
+            };
+            expect($sut)->toThrow(new InvalidStructureException());
+        });
+
+        it('should decode a valid jwt', function () {
+            $token = $this->decoder->decode($this->jwt);
+            expect($token->getClaim('custom:user_id'))->toBe('123');
+            expect($token->getClaim('token_use'))->toBe('id');
+        });
+
+        it('should accept valid claims if only one verifier succeeds', function () {
+            $invalidClaimsVerifier = Double::instance(['implements' => ClaimVerifierInterface::class]);
+        
+            allow($invalidClaimsVerifier)->toReceive('verify')->andRun(function ($token) {
+                throw new InvalidClaimsException("Those claims are not invalid!");
+            });
+
+            allow($invalidClaimsVerifier)->toReceive('getUserPool')->andReturn($this->userPool);
+
+            $decoder = new MultiPoolJwtDecoder([$invalidClaimsVerifier, $this->verifier]);
+
+            $token = $decoder->decode($this->jwt);
+            expect($token->getClaim('custom:user_id'))->toBe('123');
+            expect($token->getClaim('token_use'))->toBe('id');
+        });
+
+        it('should report the last known error produced by a claims verifier', function () {
+            $first = Double::instance(['implements' => ClaimVerifierInterface::class]);
+        
+            allow($first)->toReceive('verify')->andRun(function ($token) {
+                throw new InvalidClaimsException("Those claims are not invalid!");
```

Comment:
I know this is just a test, but we have a double negative in the exception. Also, doesn't look like we're using token.